signing/signing.go

```diff
@@ -0,0 +1,351 @@
+	if err != nil {
+		http.Error(w, "error getting repository", http.StatusInternalServerError)
+		log.Println(err, resp)
+		return
+	}
+
+	// Verify that the branch from the results files is the repo's default branch.
+	defaultBranch := "refs/heads/" + repo.GetDefaultBranch()
+	if defaultBranch != repoRef {
+		http.Error(w, "branch of cert isn't the repo's default branch", http.StatusInternalServerError)
+		log.Println(defaultBranch, repoRef)
+		return
+	}
+
+	// Get workflow file from cert commit SHA.
+	opts := &github.RepositoryContentGetOptions{Ref: repoSHA}
+	contents, _, _, err := client.Repositories.GetContents(ctx, ownerName, repoName, workflowPath, opts)
+	if err != nil {
+		http.Error(w, "error downloading workflow contents from repo", http.StatusInternalServerError)
+		log.Println(err)
+		return
+	}
+
+	workflowContent, err := contents.GetContent()
+	if err != nil {
+		http.Error(w, "error decoding workflow contents", http.StatusInternalServerError)
+		log.Println(err)
+		return
+	}
+
+	// Verify scorecard workflow.
+	err = verifyScorecardWorkflow(workflowContent)
+	if err != nil {
+		http.Error(w, "workflow could not be verified", http.StatusInternalServerError)
+		log.Println(err)
+		return
+	}
+
+	// Save scorecard results (results.sarif, results.json, score.txt) to GCS
+	bucketURL := "gs://ossf-scorecard-results"
+	folderPath := fmt.Sprintf("%s/%s", "github.com", repoPath)
+	sarifPath := fmt.Sprintf("%s/results.sarif", folderPath)
+	jsonPath := fmt.Sprintf("%s/results.json", folderPath)
+
+	err1 = data.WriteToBlobStore(ctx, bucketURL, sarifPath, []byte(scorecardOutput.SarifOutput))
+	err2 = data.WriteToBlobStore(ctx, bucketURL, jsonPath, []byte(scorecardOutput.JsonOutput))
+	if err1 != nil || err2 != nil {
+		http.Error(w, "error writing to GCS bucket", http.StatusUnauthorized)
+		log.Println(err1, err2)
+		return
+	}
+
+	// Write response.
+	w.Write([]byte(fmt.Sprintf("Successfully verified and uploaded scorecard results for repo %s on branch %s", []byte(repoName), []byte(repoRef))))
+	w.WriteHeader(http.StatusOK)
+}
+
+func lookupPayload(ctx context.Context, payload []byte) (repoPath, repoRef, repoSHA, workflowPath string, err error) {
+	// Get most recent Rekor entry uuid.
+	rekorClient, err := rekor.NewClient(options.DefaultRekorURL)
+	if err != nil {
+		return "", "", "", "", fmt.Errorf("error initializing Rekor client: %v", err)
+	}
+
+	uuids, err := cosign.FindTLogEntriesByPayload(ctx, rekorClient, []byte(payload))
+	if err != nil || len(uuids) == 0 {
+		return "", "", "", "", fmt.Errorf("error finding tlog entries corresponding to payload: %v", err)
+	}
+	uuid := uuids[len(uuids)-1] // ignore past entries.
+
+	// Get tlog entry from the UUID.
+	entry, err := cosign.GetTlogEntry(ctx, rekorClient, uuid)
+	if err != nil {
+		return "", "", "", "", fmt.Errorf("error fetching tlog entry: %v", err)
+	}
+
+	// Verify tlog entry to make sure it is actually in the log.
```

Comment:
you also need an expiry time check to make sure the signature was created when the cert was valid
https://github.com/sigstore/cosign/blob/ac682db9511cc610d5a37704776300421d2c5e30/cmd/cosign/cli/verify/verify_blob.go#L339

Reply:
added this check